[Filters] Make ImageBuffer::filteredImage() use FilterTargetSwitcher to apply the filter

shallawa opened the pull request:
#### 6b3eb0be454e99625e24e9453a48a66ae96ed6f2
<pre>
[Filters] Make ImageBuffer::filteredImage() use FilterTargetSwitcher to apply the filter
<a href="https://bugs.webkit.org/show_bug.cgi?id=248587">https://bugs.webkit.org/show_bug.cgi?id=248587</a>
rdar://102847191

Reviewed by Simon Fraser.

Filter needs to know all the FilterRenderingModes it can use when it applies its
functions/effects. Instead of holding a single FilterRenderingMode, Filter will
hold a OptionSet&lt;FilterRenderingMode&gt;. So implementing Filter::renderingMode()
will be a lot clearer.

A new overriding ImageBuffer::filteredImage() will be added to take a drawCallback
argument. This method will be used for clean ImageBuffers which has no drawings
yet. If FilterRenderingMode::GraphicsContext can be used, the styles transparency
layers will be begun before calling drawCallback. And they will be ended after
calling drawCallback.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::filteredImage):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::renderingMode const):
(WebCore::Filter::setFilterRenderingModes):
(WebCore::Filter::setFilterRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterRenderingModes const):
(WebCore::Filter::filterRenderingMode const): Deleted.
(WebCore::Filter::setFilterRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createApplier const):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
(WebCore::FilterStyleTargetSwitcher::FilterStyleTargetSwitcher):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):

Canonical link: <a href="https://commits.webkit.org/257275@main">https://commits.webkit.org/257275@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2870b2d018fb8ae33914fdfe8e42e2b81497d8f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107701 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167982 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102190 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7958 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36219 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104285 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103902 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5980 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84841 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33080 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89549 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76024 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1418 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1372 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22484 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5015 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6255 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44989 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41909 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ImageBuffer.h:
```diff
@@ -171,6 +171,8 @@ class ImageBuffer : public ThreadSafeRefCounted<ImageBuffer>, public CanMakeWeak
     WEBCORE_EXPORT virtual RefPtr<NativeImage> sinkIntoNativeImage();
     WEBCORE_EXPORT RefPtr<Image> copyImage(BackingStoreCopy = CopyBackingStore, PreserveResolution = PreserveResolution::No) const;
     WEBCORE_EXPORT virtual RefPtr<Image> filteredImage(Filter&);
+    RefPtr<Image> filteredImage(Filter&, std::function<void(GraphicsContext&)> drawCallback);
```
Can the Filter& be const?

shallawa replied:
Yes it can. I will clean this issue in a separate patch.

shallawa committed:
[Filters] Make ImageBuffer::filteredImage() use FilterTargetSwitcher to apply the filter
https://bugs.webkit.org/show_bug.cgi?id=248587
rdar://102847191

Reviewed by Simon Fraser.

Filter needs to know all the FilterRenderingModes it can use when it applies its
functions/effects. Instead of holding a single FilterRenderingMode, Filter will
hold a OptionSet<FilterRenderingMode>. So implementing Filter::renderingMode()
will be a lot clearer.

A new overriding ImageBuffer::filteredImage() will be added to take a drawCallback
argument. This method will be used for clean ImageBuffers which has no drawings
yet. If FilterRenderingMode::GraphicsContext can be used, the styles transparency
layers will be begun before calling drawCallback. And they will be ended after
calling drawCallback.

* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::filteredImage):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::renderingMode const):
(WebCore::Filter::setFilterRenderingModes):
(WebCore::Filter::setFilterRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterRenderingModes const):
(WebCore::Filter::filterRenderingMode const): Deleted.
(WebCore::Filter::setFilterRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createApplier const):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
(WebCore::FilterStyleTargetSwitcher::FilterStyleTargetSwitcher):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h:
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):

Canonical link: https://commits.webkit.org/257275@main
